extensions/amp-story/1.0/amp-story.js

```diff
@@ -1284,17 +1328,53 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
-    if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
-    ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
-        return Promise.resolve();
-      }
-      this.storeService_.dispatch(Action.TOGGLE_SUBSCRIPTIONS_DIALOG, true);
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
+    if (this.isPaywallStory_()) {
+      if (
+        pageIndex >= PAYWALL_PAGE_INDEX &&
+        subscriptionsState !== SubscriptionsState.GRANTED
+      ) {
+        if (subscriptionsState === SubscriptionsState.UNKNOWN) {
+          return this.blockOnPendingSubscriptionsState_(
+            targetPageId,
+            direction
+          );
+        }
+
+        // Attempt to navigate to the locked pages after the paywall page.
+        if (pageIndex > PAYWALL_PAGE_INDEX) {
+          this.pageAfterGranted_ = targetPageId;
 
-      // TODO(#37285): add SubscriptionService to actually trigger the subscription dialog.
+          // Show the paywall immediately if the active page is the paywall page.
+          if (this.isOnPaywallPage_()) {
+            this.showSubscriptionsDialog_();
+            return Promise.resolve();
+          }
+
+          // Redirect to the paywall page.
+          return this.switchTo_(
+            this.pages_[PAYWALL_PAGE_INDEX].element.id,
+            direction
+          );
```

Comment:
`switchTo` will early return if you're trying to navigate to the currently active page, so you can simplify your code a lot:

```suggestion
          // Redirects to the last authorized page if needed, and shows paywall.
          this.showSubscriptionsDialog_();
          return this.switchTo_(
            this.pages_[PAYWALL_PAGE_INDEX].element.id,
            direction
          );
```

Reply:
We would still need that check to do 
```
if (this.isOnPaywallPage_()) {
    this.showSubscriptionsDialog_();
}
```
because we only show the paywall on paywall page.

Reply:
Well if you're calling the code to navigate to the paywall page you'll always be on the paywall page?

Reply:
FYI separating out the logic of showing the UI and switch to the paywall page does remove the need of the check : )